### PR TITLE
Scan i2c bus command

### LINKED ISSUE #2: Add I2C scan telecommands
Reference the I2C scanner example in the `debug_tools/debug_i2c.c` file.

* Argument 1: Which I2C port to scan, as an integer

### PR DESCRIPTION
**Pull request by parkerg-13**
Fixes #2

### CODE REVIEW

File: firmware/Core/Inc/telecommands/i2c_telecommand_defs.h [hunk 1]
```diff
@@ -0,0 +1,7 @@
+#include <stdint.h>
+#include "telecommands/telecommand_definitions.h"
+#include "main.h"
```
Comment (DeflateAwning) [inline firmware/Core/Inc/telecommands/i2c_telecommand_defs.h:3]: This file needs an include guard (see other header files for examples).

Please include other header files from the `.c` files, and not from `.h` files.

File: firmware/Core/Inc/telecommands/i2c_telecommand_defs.h [hunk 2]
```diff
@@ -0,0 +1,7 @@
...
+
+
+uint8_t TCMDEXEC_scan_i2c_device(const char *args_str, TCMD_TelecommandChannel_enum_t tcmd_channel,
```
Comment (DeflateAwning) [inline firmware/Core/Inc/telecommands/i2c_telecommand_defs.h:6]: Can we rename from `scan_i2c_device` to `scan_i2c_bus`? We're not scanning a single device, but rather scanning for any/all devices on a bus.

File: firmware/Core/Src/telecommands/i2c_telecommand_defs.c [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+#include <stdio.h>
+#include <stdlib.h>
+
+uint8_t TCMDEXEC_scan_i2c_device(const char *args_str, TCMD_TelecommandChannel_enum_t tcmd_channel,
+                        char *response_output_buf, uint16_t response_output_buf_len) {
+
+    snprintf(response_output_buf, response_output_buf_len, "Starting I2C scan...\n");
+
+    I2C_HandleTypeDef* hi2c;
+    int bus_to_scan = atoi(args_str);
```
Comment (DeflateAwning) [inline firmware/Core/Src/telecommands/i2c_telecommand_defs.c:12]: What happens if this fails? Also, please use `uint8_t` (or similar specific type) instead of `int`

File: firmware/Core/Src/telecommands/i2c_telecommand_defs.c [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+            break;
+        case 3: 
+            hi2c = &hi2c3;
+            break;
+        case 4:
+            hi2c = &hi2c4;
+            break;
+        default:
+            hi2c = NULL;
+            snprintf(response_output_buf, response_output_buf_len, "Enter a valid I2C bus!\n");
```
Comment (DeflateAwning) [inline firmware/Core/Src/telecommands/i2c_telecommand_defs.c:30]: This over-writes the "Starting I2C scan..." message. Perhaps we want to remove the "Starting scan" message?

File: firmware/Core/Src/telecommands/i2c_telecommand_defs.c [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+            return 0;
+    }
+
+
+	char msg[5];
+	// Go through all possible i2c addresses
+	for (uint8_t i = 0; i < 128; i++) {
+
+		if (HAL_I2C_IsDeviceReady(hi2c, (uint16_t)(i<<1), 3, 5) == HAL_OK) {
```
Comment (DeflateAwning) [inline firmware/Core/Src/telecommands/i2c_telecommand_defs.c:39]: What are the `3, 5` args? Looks like magic numbers, which are bad.

File: firmware/Core/Src/telecommands/i2c_telecommand_defs.c [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+			// We got an ack
+			sprintf(msg, "%2x ", i);
+			snprintf(response_output_buf, response_output_buf_len, "%s", msg);
```
Comment (DeflateAwning) [inline firmware/Core/Src/telecommands/i2c_telecommand_defs.c:42]: Doesn't this overwrite at the start?

File: firmware/Core/Src/telecommands/telecommand_definitions.c [hunk 7]
```diff
@@ -233,5 +239,4 @@ uint8_t TCMDEXEC_available_telecommands(const char *args_str, TCMD_TelecommandCh
     snprintf(response_output_buf, response_output_buf_len, "%s", response);
 
     return 0;
-}
-
+}
```
Comment (DeflateAwning) [inline firmware/Core/Src/telecommands/telecommand_definitions.c:242]: Avoid unnecessary whitespace changes

**Review by DeflateAwning — CHANGES_REQUESTED** [review]
Very solid start! Looks like there are some logical bugs and some "best practice" opportunities.

Please ensure you've tested code before you request a review. If you test this without any I2C devices connected, the output should represent that there are no I2C devices connected.

File: firmware/Core/Src/telecommands/i2c_telecommand_defs.c (hunk 5 above)
Comment (DeflateAwning) [reply firmware/Core/Src/telecommands/i2c_telecommand_defs.c:39]: Here, can we also please track the different return statuses from `HAL_I2C_IsDeviceReady` so that `HAL_TIMEOUT` is represented differently than `HAL_ERROR` ?

File: firmware/Core/Inc/telecommands/i2c_telecommand_defs.h (hunk 1 above)
Comment (parkerg-13) [reply firmware/Core/Inc/telecommands/i2c_telecommand_defs.h:3]: Done.

File: firmware/Core/Inc/telecommands/i2c_telecommand_defs.h (hunk 2 above)
Comment (parkerg-13) [reply firmware/Core/Inc/telecommands/i2c_telecommand_defs.h:6]: Done.

File: firmware/Core/Src/telecommands/i2c_telecommand_defs.c (hunk 3 above)
Comment (parkerg-13) [reply firmware/Core/Src/telecommands/i2c_telecommand_defs.c:14]: If the conversion fails atoi returns 0 which is handled by the default case in the switch statement. Should this be handled explicitly before we hit the switch statement?

File: firmware/Core/Src/telecommands/i2c_telecommand_defs.c (hunk 5 above)
Comment (parkerg-13) [reply firmware/Core/Src/telecommands/i2c_telecommand_defs.c:40]: In reference to the 3, 5 args I can give them variable names. Are these numbers what we want? The 3 represents the number of trials to see if the device is ready and the 5 represents the timeout in milliseconds.

File: firmware/Core/Src/telecommands/i2c_telecommand_defs.c (hunk 4 above)
Comment (parkerg-13) [reply firmware/Core/Src/telecommands/i2c_telecommand_defs.c:32]: I misunderstood the functioning of this way of writing to UART. Fix will be soon.

File: firmware/Core/Src/telecommands/i2c_telecommand_defs.c (hunk 6 above)
Comment (parkerg-13) [reply firmware/Core/Src/telecommands/i2c_telecommand_defs.c:43]: Yes misunderstood how things were working this has been fixed.

File: firmware/Core/Src/telecommands/i2c_telecommand_defs.c (hunk 5 above)
Comment (parkerg-13) [reply firmware/Core/Src/telecommands/i2c_telecommand_defs.c:40]: Error messages separated.

**Comment by parkerg-13** [comment]
Still have to test with devices attached.

File: firmware/Core/Inc/rtos_tasks/rtos_tasks.h [hunk 8]
```diff
@@ -5,6 +5,8 @@
 #include "main.h"
 #include "rtos_tasks/rtos_task_helpers.h"
 
+#define MAX_RESPONSE_BUFFER_LENGTH 1024
```
Comment (DeflateAwning) [inline firmware/Core/Inc/rtos_tasks/rtos_tasks.h:8]: Requires namespace prefix (probably `TCMD_`)

File: firmware/Core/Src/telecommands/i2c_telecommand_defs.c [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+    HAL_StatusTypeDef i2c_device_status;
+
+	// Go through all possible i2c addresses
+	for (uint8_t i = 0; i < 128; i++) {
+
+        i2c_device_status = HAL_I2C_IsDeviceReady(hi2c, (uint16_t)(i<<1), 3, 5);
+
+		if (i2c_device_status == HAL_OK) {
+			// O for OK
+			sprintf(msg, "%2x O\n", i);
```
Comment (DeflateAwning) [inline firmware/Core/Src/telecommands/i2c_telecommand_defs.c:43]: Please use `snprintf` for length overflow safety. No `snprintf`. `strcat` is also unsafe.

File: firmware/Core/Src/telecommands/telecommand_definitions.c [hunk 10]
```diff
@@ -9,6 +9,7 @@
 #include "telecommands/flash_telecommand_defs.h"
 #include "telecommands/lfs_telecommand_defs.h"
 #include "telecommands/timekeeping_telecommand_defs.h"
+#include "i2c_telecommand_defs.h"
```
Comment (DeflateAwning) [inline firmware/Core/Src/telecommands/telecommand_definitions.c:12]: Please add folder prefix

File: firmware/Core/Src/telecommands/i2c_telecommand_defs.c [hunk 11]
```diff
@@ -0,0 +1,68 @@
...
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include "main.h"
+
+uint8_t TCMDEXEC_scan_i2c_bus(const char *args_str, TCMD_TelecommandChannel_enum_t tcmd_channel,
+                        char *response_output_buf, uint16_t response_output_buf_len) {
+
+    I2C_HandleTypeDef* hi2c;
+    uint8_t bus_to_scan = atoi(args_str);
```
Comment (DeflateAwning) [inline firmware/Core/Src/telecommands/i2c_telecommand_defs.c:11]: What happens if `args_str` isn't valid? I don't think that's handled here currently

**Review by DeflateAwning — CHANGES_REQUESTED** [review]
More comments. Good changes overall though

File: firmware/Core/Inc/rtos_tasks/rtos_tasks.h (hunk 8 above)
Comment (parkerg-13) [reply firmware/Core/Inc/rtos_tasks/rtos_tasks.h:8]: Done.

File: firmware/Core/Src/telecommands/i2c_telecommand_defs.c (hunk 9 above)
Comment (parkerg-13) [reply firmware/Core/Src/telecommands/i2c_telecommand_defs.c:43]: Done.

File: firmware/Core/Src/telecommands/telecommand_definitions.c (hunk 10 above)
Comment (parkerg-13) [reply firmware/Core/Src/telecommands/telecommand_definitions.c:12]: Done.

File: firmware/Core/Src/telecommands/i2c_telecommand_defs.c (hunk 11 above)
Comment (parkerg-13) [reply firmware/Core/Src/telecommands/i2c_telecommand_defs.c:11]: What do you mean by not valid? If anything funky is passed to atoi then it will just return 0 and we will handle it as something going wrong in the switch statement.

**Comment by DeflateAwning** [comment]
Changes look good. Can you please rebase to `main`?

File: firmware/Core/Src/telecommands/i2c_telecommand_defs.c (hunk 11 above)
Comment (DeflateAwning) [reply firmware/Core/Src/telecommands/i2c_telecommand_defs.c:11]: For example, if `args_str = "this is not an integer"`, it's important we consider what happens. Because there's no "i2c 0", that works well enough here.

**Review by DeflateAwning — APPROVED** [review]
Great work, Parker G! 

I rebased to `main` so that this can be merged. I also added counts of each status to the end of the message. 

Feel free to add another commit in this branch to add your name to the list of contributors in the README! When you're ready, you can click "Squash and Merge" to merge this change into `main`.

**Review by DeflateAwning — DISMISSED** [review]

**Review by DeflateAwning — APPROVED** [review]
